tsdb/engine/tsm1/engine.go

```diff
@@ -2498,6 +2498,40 @@ type indexTagSets interface {
 	TagSets(name []byte, options query.IteratorOptions) ([]*query.TagSet, error)
 }
 
+// createSeriesIterator creates an optimized series iterator if possible.
+// We exclude less-common cases for now as not worth implementing.
+func (e *Engine) createSeriesIterator(measurement string, ref *influxql.VarRef, is tsdb.IndexSet, opt query.IteratorOptions) (query.Iterator, error) {
+	// Main check to see if we are trying to create a seriesKey iterator
+	if ref == nil || ref.Val != "_seriesKey" || len(opt.Aux) != 0 {
+		return nil, nil
+	}
+	// Check some other cases that we could maybe handle, but don't
+	if len(opt.Dimensions) > 0 {
+		return nil, nil
+	}
+	if opt.SLimit != 0 || opt.SOffset != 0 {
+		return nil, nil
+	}
+	if opt.StripName {
+		return nil, nil
+	}
+	if opt.Ordered {
+		return nil, nil
+	}
+	// Actual creation of the iterator
+	seriesCursor, err := is.MeasurementSeriesKeyByExprIterator([]byte(measurement), opt.Condition, opt.Authorizer)
+	if err != nil {
+		seriesCursor.Close()
```

Comment:
Can the` Close()` cause problems? Are there errors where the `seriesCursor` is nil, or already closed or in any other state where `Close()` would fail or panic? We have the problem elsewhere.

Reply:
Looking at the implementation for MeasurementSeriesKeyByExprIterator, the contract is it either returns `nil, error` or `cursor, nil`, never both.